aten/src/ATen/native/TensorCompare.cpp

```diff
@@ -106,6 +108,44 @@ Tensor isinf(const Tensor &self) {
   });
 }
 
+static inline TensorIterator build_is_infinity_op_iterator(Tensor& result, const Tensor& self) {
+  return at::TensorIteratorConfig()
+    .check_all_same_dtype(false)
+    .set_check_mem_overlap(true)
+    .add_output(result)
+    .add_input(self)
+    .cast_common_dtype_to_outputs(true)
+    .build();
+}
+
+Tensor isposinf(const Tensor &self) {
+  Tensor result = at::empty_like(self, at::kBool, at::MemoryFormat::Preserve);
+  at::isposinf_out(result, self);
+  return result;
+}
+
+Tensor& isposinf_out(Tensor& result, const Tensor& self) {
+  TORCH_CHECK(!self.is_complex(), "isposinf does not support complex inputs.");
+  result.resize_(self.sizes());
+  auto iter = build_is_infinity_op_iterator(result, self);
+  isposinf_stub(iter.device_type(), iter);
+  return result;
+}
+
+Tensor isneginf(const Tensor &self) {
+  Tensor result = at::empty_like(self, at::kBool, at::MemoryFormat::Preserve);
+  at::isneginf_out(result, self);
+  return result;
+}
+
+Tensor& isneginf_out(Tensor& result, const Tensor& self) {
+  TORCH_CHECK(!self.is_complex(), "isneginf does not support complex inputs.");
```

Comment:
Better add a check for result being a bool tensor here, too

Reply:
On lines 151 and 152 below, in the `isfinite` function, there's a special case for handling bool and integer tensors:

```
if (c10::isIntegralType(self.scalar_type(), /*include_bool=*/true)) {
  return at::ones_like(self, at::kBool, at::MemoryFormat::Preserve);	   
}
```

I suggest you adopt these special case in your functions, too, for simplicity. For the out version I think you can use `result.fill_(false)`, instead. 

This way your TensorIterator will only have to deal with float types.

Let me know if this makes sense.

Reply:
Thank you so much for your advice.
I have adopted this approach and the latest PR only uses TensorIterator to deal with float types, the kernels are simplified greatly.